File: yocs_navigator/src/lib/basic_move_controller.cpp
/*
  Basic Move Controller - Provides simple robot behavior such as go forward backward

  LICENSE : BSD - https://raw.github.com/yujinrobot/yujin_ocs/license/LICENSE
 */
#include "yocs_navigator/basic_move_controller.hpp" 

namespace yocs_navigator {

BasicMoveController::BasicMoveController(ros::NodeHandle& n) 
: nh_(n), cmd_vel_topic_(BasicMoveControllerDefaultParam::PUB_CMD_VEL), odometry_topic_(BasicMoveControllerDefaultParam::SUB_ODOM)
{
  init();
}

BasicMoveController::BasicMoveController(ros::NodeHandle& n, const std::string& cmd_vel_topic, const std::string& odometry_topic) 
: nh_(n), cmd_vel_topic_(cmd_vel_topic), odometry_topic_(odometry_topic)
{
  init();
}

BasicMoveController::~BasicMoveController()
{
}

void BasicMoveController::init() 
{
  pub_cmd_vel_ = nh_.advertise<geometry_msgs::Twist>(cmd_vel_topic_, 5);
  sub_odom_    = nh_.subscribe(odometry_topic_, 1, &BasicMoveController::processOdometry, this);
}

void BasicMoveController::processOdometry(const nav_msgs::Odometry::ConstPtr& msg)
{   
  odometry_ = *msg;
}   

void BasicMoveController::slowForward()
{
  moveAt( 0.1,  0.0,  0.1);
}

void BasicMoveController::slowBackward()
{
  moveAt(-0.1,  0.0,  0.1);
}

void BasicMoveController::turnClockwise()
{
  moveAt( 0.0, -0.5,  0.1);
}

void BasicMoveController::turnCounterClockwise()
{
  moveAt( 0.0,  0.5,  0.1);
}

void BasicMoveController::stop()
{
  moveAt( 0.0,  0.0,  0.0);
}

void BasicMoveController::moveAt(double v, double w, double t = 0.0)
{
  geometry_msgs::Twist vel;
  vel.linear.x  = v;
  vel.angular.z = w;
  pub_cmd_vel_.publish(vel);
  ros::Duration(t).sleep();
}

void BasicMoveController::forward(double distance)
{
  geometry_msgs::Point pos0 = odometry_.pose.pose.position;
  while (mtk::distance2D(pos0, odometry_.pose.pose.position) < distance)
  {
    slowForward();
  }
}

void BasicMoveController::backward(double distance)
{
  geometry_msgs::Point pos0 = odometry_.pose.pose.position;
  while(mtk::distance2D(pos0, odometry_.pose.pose.position) < distance)
  {
    slowBackward();
  }
}

void BasicMoveController::turn(double angle)
{
  double yaw0 = tf::getYaw(odometry_.pose.pose.orientation);
  double yaw1 = mtk::wrapAngle(yaw0 + angle);

  ROS_DEBUG("%f  %f  %f", angle, yaw0, yaw1);
  while (std::abs(mtk::wrapAngle(yaw1 - tf::getYaw(odometry_.pose.pose.orientation))) > 0.05)
    moveAt(0.0, mtk::sign(angle)*0.5, 0.05);
}

// aaaagh...  TODO make a decent version checking sign change and turns over 2pi! I have no time now
void BasicMoveController::turn2(double angle)
{
  double yaw0 = tf::getYaw(odometry_.pose.pose.orientation);
  double yaw1 = mtk::wrapAngle(yaw0 + angle);
  double sign = mtk::sign(yaw1 - yaw0);

  ROS_DEBUG("%f  %f  %f", angle, yaw0, yaw1);
  while (mtk::sign(mtk::wrapAngle(tf::getYaw(odometry_.pose.pose.orientation) - yaw1)) == sign)
  {
    moveAt(0.0, mtk::sign(angle)*0.5, 0.05);
  }
}

void BasicMoveController::spinClockwise()
{
  // WARN; note that this requires that callbacks keep running!
  double yaw0 = tf::getYaw(odometry_.pose.pose.orientation);

  for (int i = 0; i < 5 || tf::getYaw(odometry_.pose.pose.orientation) <= yaw0; ++i)
    turnClockwise();

  for (int i = 0; i < 5 || tf::getYaw(odometry_.pose.pose.orientation) >  yaw0; ++i)
    turnClockwise();
}

void BasicMoveController::spinCounterClockwise()
{
  // WARN; note that this requires that callbacks keep running!
  double yaw0 = tf::getYaw(odometry_.pose.pose.orientation);

  for (int i = 0; i < 5 || tf::getYaw(odometry_.pose.pose.orientation) >= yaw0; ++i)
    turnCounterClockwise();

  for (int i = 0; i < 5 || tf::getYaw(odometry_.pose.pose.orientation) <  yaw0; ++i)
    turnCounterClockwise();
}

} // namespace yocs

File: yocs_navigator/src/yocs_navigator/basic_move_controller.py
'''
 basic_move_controller.py
 LICENSE : BSD - https://raw.github.com/yujinrobot/yujin_ocs/license/LICENSE
'''
import math
import copy
import rospy
import tf
import nav_msgs.msg as nav_msgs
import geometry_msgs.msg as geometry_msgs

class BasicMoveController(object):
    def __init__(self, odom_topic='odom', cmd_vel_topic='cmd_vel'):
        self._sub_odom = rospy.Subscriber(odom_topic, nav_msgs.Odometry, self.process_odometry)
        self._pub_cmd_vel = rospy.Publisher(cmd_vel_topic, geometry_msgs.Twist, queue_size=5)
        
        self._odom = None

    def process_odometry(self, msg):
        self._odom = copy.deepcopy(msg)

    def move_at(self, v, w, t):
        vel = geometry_msgs.Twist()

        vel.linear.x = v
        vel.angular.z = w
        self._pub_cmd_vel.publish(vel)
        rospy.sleep(t)

    def turn_clockwise(self):
        self.move_at(0.0, -0.5, 0.1)

    def turn_counter_clockwise(self):
        self.move_at(0.0, 0.5, 0.1)

    def turn(self, angle):
        raise NotImplementedError()

    def spin_clockwise(self):
        yaw = self._get_odom_yaw()

        i = 0
        while (self._get_odom_yaw() <= yaw) or i < 5:
            self.turn_clockwise()
            i = i + 1

        i = 0
        while (self._get_odom_yaw() > yaw) or i < 5:
            self.turn_clockwise()
            i = i +1

    def _get_odom_yaw(self):
        quaternion = (self._odom.pose.pose.orientation.x, self._odom.pose.pose.orientation.y, self._odom.pose.pose.orientation.z, self._odom.pose.pose.orientation.w)
        roll, pitch, yaw = tf.transformations.euler_from_quaternion(quaternion)
        return yaw

    def slow_forward(self):
        self.move_at(0.1, 0.0, 0.1)

    def slow_backward(self):
        self.move_at(-0.1, 0.0, 0.1)

    def forward(self, distance):
        pos0 = copy.deepcopy(self._odom.pose.pose.position)

        while(self._distance2d(pos0, self._odom.pose.pose.position) < distance):
            self.slow_forward()

    def backward(self, distance):
        pos0 = copy.deepcopy(self._odom.pose.pose.position)

        while(self._distance2d(pos0, self._odom.pose.pose.position) < distance):
            self.slow_backward()

    def _distance2d(self, p1, p2):
        '''
        computes distance between two points
        '''
        return math.sqrt(math.pow((p2.x - p1.x), 2) + math.pow((p2.y -p1.y), 2))
